ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryConsistencyFunctionalTest.java

```diff
@@ -0,0 +1,460 @@
+    ActiveStandbyEntity entity1 = clusterStatusResponse.getClusterStatus().get(testApp1.getHost())
+      .getActiveStandbyPerQuery().get(queryId);
+
+    // find active
+    if (!entity0.getActiveStores().isEmpty() && !entity0.getActivePartitions().isEmpty()) {
+      clusterFormation.setActive(testApp0);
+    } else if (!entity1.getActiveStores().isEmpty() && !entity1.getActivePartitions().isEmpty()) {
+      clusterFormation.setActive(testApp1);
+    } else {
+      clusterFormation.setActive(testApp2);
+    }
+
+    //find standby
+    if (!entity0.getStandByStores().isEmpty() && !entity0.getStandByPartitions().isEmpty()) {
+      clusterFormation.setStandBy(testApp0);
+    } else if (!entity1.getStandByStores().isEmpty() && !entity1.getStandByPartitions().isEmpty()) {
+      clusterFormation.setStandBy(testApp1);
+    } else {
+      clusterFormation.setStandBy(testApp2);
+    }
+
+    //find router
+    if (entity0.getStandByStores().isEmpty() && entity0.getActiveStores().isEmpty()) {
+      clusterFormation.setRouter(testApp0);
+    } else if (entity1.getStandByStores().isEmpty() && entity1.getActiveStores().isEmpty()) {
+      clusterFormation.setRouter(testApp1);
+    } else {
+      clusterFormation.setRouter(testApp2);
+    }
+
+    return clusterFormation;
+  }
+
+  static class ClusterFormation {
+    TestApp active;
+    TestApp standBy;
+    TestApp router;
+
+    ClusterFormation() {
+    }
+
+    public void setActive(final TestApp active) {
+      this.active = active;
+    }
+
+    public void setStandBy(final TestApp standBy) {
+      this.standBy = standBy;
+    }
+
+    public void setRouter(final TestApp router) {
+      this.router = router;
+    }
+
+    public String toString() {
+      return new StringBuilder()
+        .append("Active = ").append(active.getHost())
+        .append(", Standby = ").append(standBy.getHost())
+        .append(", Router = ").append(router.getHost())
+        .toString();
+    }
+  }
+
+  private void waitForTableRows() {
+    TEST_HARNESS.verifyAvailableUniqueRows(
+      output.toUpperCase(),
+      USER_PROVIDER.data().size(),
+      FormatFactory.KAFKA,
+      FormatFactory.JSON,
+      AGGREGATE_SCHEMA
+    );
+  }
+
+  private static String getNewStateDir() {
+    try {
+      return TMP.newFolder().getAbsolutePath();
+    } catch (final IOException e) {
+      throw new AssertionError("Failed to create new state dir", e);
+    }
+  }
+
+  private static class TestApp {
+
+    private final KsqlHostInfoEntity host;
+    private final TestKsqlRestApp app;
+    private final Shutoffs shutoffs;
+
+    public TestApp(KsqlHostInfoEntity host, TestKsqlRestApp app, Shutoffs shutoffs) {
+      this.host = host;
+      this.app = app;
+      this.shutoffs = shutoffs;
+    }
+
+    public KsqlHostInfoEntity getHost() {
+      return host;
+    }
+
+    public TestKsqlRestApp getApp() {
+      return app;
+    }
+
+    public Shutoffs getShutoffs() {
```

Comment:
This seems not used?

Reply:
Good catch @guozhangwang . The thing is, this test class is where lots of the consistency tests will eventually go in. Currently is just checking that forwarding and returning of the token works, but eventually we want to test consistency tokens are handled as expected, also in the presence of shutdown/rebalance/etc, so I prefer to keep those infra calls for the moment.